src/mito2/src/region/opener.rs

```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
             }
         );
 
+        let region_id = metadata.region_id;
         let mutable = self.memtable_builder.build(&metadata);
         let version = VersionBuilder::new(metadata, mutable).build();
-        let version_control = Arc::new(VersionControl::new(version));
+        let flushed_sequence = version.flushed_sequence;
 
-        // TODO(yingwen): Replay.
+        let version_control = Arc::new(VersionControl::new(version));
+        replay_memtable(
+            &self.wal,
+            region_id,
+            flushed_sequence,
+            version_control.clone(),
+        )
+        .await?;
 
-        Ok(MitoRegion {
+        let region = MitoRegion {
             region_id: self.region_id,
             version_control,
             manifest_manager,
-        })
+        };
+        Ok(region)
     }
 }
 
+/// Replays the mutations from WAL and inserts mutations to memtable of given region.
+async fn replay_memtable<S: LogStore>(
+    wal: &Wal<S>,
+    region_id: RegionId,
+    flushed_sequence: SequenceNumber,
+    version_control: VersionControlRef,
+) -> Result<()> {
+    let mut notify_waiter = vec![];
+    {
+        let mut region_writer_ctx = RegionWriteCtx::new(region_id, version_control);
+        let mut wal_stream = wal.scan(region_id, flushed_sequence)?;
+        while let Some(res) = wal_stream.next().await {
+            let (_, entry) = res?;
```

Comment:
We might need to store `flushed_entry_id` and use it instead of `flushed_sequence`.

We must also initialize `last_entry_id` and `committed_sequence` to update them to correct values after `write_memtables()`. I updated the entry id and committed sequence in https://github.com/GreptimeTeam/greptimedb/pull/2230

Reply:
Maybe add a todo here?

Reply:
Maybe a FIXME